src/components/organisms/SchedulePageModal/SchedulePageModal.utils.ts

```diff
@@ -0,0 +1,97 @@
+import {
+  addDays,
+  compareAsc,
+  differenceInMinutes,
+  endOfDay,
+  format,
+  isWithinInterval,
+  startOfDay,
+} from "date-fns";
+
+import { Room } from "types/rooms";
+import { PersonalizedVenueEvent, VenueEvent } from "types/venues";
+import { MyPersonalizedSchedule } from "types/User";
+
+import { WithVenueId } from "utils/id";
+import { ONE_MINUTE_IN_SECONDS, ONE_SECOND_IN_MILLISECONDS } from "utils/time";
+import { isTruthy } from "utils/types";
+
+import { RoomWithEvents } from "components/molecules/Schedule/Schedule.types";
+
+export const isEventThisDay = (date: Date) => {
+  return (event: VenueEvent) => {
+    return isWithinInterval(date, {
+      start: startOfDay(event.start_utc_seconds * ONE_SECOND_IN_MILLISECONDS),
+      end: endOfDay(
+        (event.start_utc_seconds +
+          event.duration_minutes * ONE_MINUTE_IN_SECONDS) *
+          ONE_SECOND_IN_MILLISECONDS
+      ),
+    });
+  };
+};
+
+export const adjustEventTiming = (date: Date) => {
+  return (event: WithVenueId<VenueEvent>) => {
+    const dates = [
+      event.start_utc_seconds * ONE_SECOND_IN_MILLISECONDS,
+      (event.start_utc_seconds +
+        event.duration_minutes * ONE_MINUTE_IN_SECONDS) *
+        ONE_SECOND_IN_MILLISECONDS,
+      startOfDay(date),
+      endOfDay(date),
+    ].sort(compareAsc);
+
+    const startUtcMilliseconds =
+      dates[1] instanceof Date ? dates[1].getTime() : dates[1];
+
+    return {
+      ...event,
+      start_utc_seconds: startUtcMilliseconds / ONE_SECOND_IN_MILLISECONDS,
+      duration_minutes: differenceInMinutes(dates[2], dates[1]),
+    };
+  };
+};
+
+export const extendRoomsWithDaysEvents = (
+  rooms: Room[],
+  daysEvents: WithVenueId<PersonalizedVenueEvent>[]
+): RoomWithEvents[] => {
+  return rooms.map((room) => {
+    const events: WithVenueId<PersonalizedVenueEvent>[] = daysEvents.filter(
+      (event) => event?.room === room?.title
+    );
+    return { ...room, events };
+  });
+};
```

Comment:
I think the logic here is going to be fairly inefficient (filtering `daysEvents` for every single room)

I wonder if we can refactor it somehow, eg. extracting all of the room titles we need, then filtering/reducing daysEvents once based on that.

Reply:
okay, will do it tomorrow morning. (haven't noticed this comment)

Reply:
filtering rooms first, and them extending them with events.
https://github.com/sparkletown/sparkle/pull/1302/commits/5461bb17ff72bc8a2ddc452452a2b5fe35d2376d